doc/developer/design/20230227_external_introspection.md

```diff
@@ -0,0 +1,175 @@
+operators in a dataflow, including the transfer of data along channels
+and the number of records in arrangements.
+
+This will differ from the presently-available GUIs in the following
+ways:
+
+* It will be externally visible
+* The user will be able to dynamically refine the zoom level by
+  clicking (rather than having to scroll down as in the current
+  hierarchical UI)
+* We will include scheduling durations alongside the other per-node
+  information (arrangement sizes / channel traffic)
+* If possible (TBD) we should figure out how to filter out the error
+  paths.
+
+Mockup: ![diagram of proposed feature](hdv.png)
+
+## Global frontier lag visualizer
+
+The global frontier lag visualizer will allow users to see at a glance
+whether dataflows are falling behind. It will display the controller's
+view of each source and export frontier, and color nodes if their
+outputs lag significantly behind their inputs.
+
+There may be multiple values corresponding to a single export, if
+the cluster on which the source dataflow is running has more than one
+replica. Each edge between dataflows will be labeled with the maximum
+frontier across all replicas.
+
+The view will incorporate all possible inputs and outputs of a
+dataflow: indexes, MVs, sources, and sinks.
+
+Mockup: ![diagram of proposed feature](gflv.png)
+
+## "Flamegraph" views of resource usage information
+
+By "resource usage information" we mean the same things that are displayed
+per-node in the hierarchical dataflow visualizer.
+
+The scope structure should let us compute various values (time spent,
+records contained) and display them in a tree-like visualization, as
+shown below.
+
+![diagram of proposed feature](fg.png)
+
+## Per-query metadata
+
+We should collected the following metadata for each query:
+
+* The timestamp at which the query executed
+* The frontiers of all dependencies
+* The optimized and physical query plans
+* The ID of the dataflow (if it is not a simple peek) used to service
+  the query
+* The SQL text of the query
+
+We can then save all this information in a table with non-trivial
+retention, and surface it from the web UI.
+
+## User-friendly rendering of query plans
+
+Currently we have text-only `EXPLAIN PLAN` output. We should render
+this data in visual form, as a graph. We should also (whenever
+possible) flow column names from the source relations through the
+nodes, so that we can show something more useful than `#0`, etc.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+## Data sources
+All features will be implemented in the web UI using React, querying
+the `mz_internal` relations for the necessary data. Relations that
+will be used include:
+
+* For the hierarchical dataflow visualizer:
+    * `mz_dataflow_operators`
+    * `mz_dataflow_channels`
+    * `mz_arrangement_sizes`
+    * `mz_dataflow_addresses`
+* For the global frontier lag visualizer:
+    * `mz_object_dependencies`
```

Comment:
The problem is that `mz_worker_compute_dependencies` is per-replica, so it's quite annoying to stitch it together for all replicas (and impossible to do in one query). A more fundamental issue is that if the replica is unhealthy, you may not be able to query the per-replica sources at all.

I was thinking we would write a query to filter out the things we don't care about in `mz_object_dependencies` (e.g. views and types).

However, your point about `SUBSCRIBE` not being present is true. Maybe we need to start collecting subscribe dependencies in a table.

Reply:
> The problem is that mz_worker_compute_dependencies is per-replica, so it's quite annoying to stitch it together for all replicas (and impossible to do in one query). A more fundamental issue is that if the replica is unhealthy, you may not be able to query the per-replica sources at all.

Since all replicas are running the same dataflows, the contents of that relation should also be the same, unless we have bugs. In contrast to operator logging, dependency information is only removed when clusterd receives an `AllowCompaction([])` command, which only happens when the adapter has told the controller to drop the dataflow, so that information stays in the logging sources of all replicas as long as the corresponding catalog object exists. So we should be fine to query any replica for dependency information.

It is true that this stops working with unhealthy replicas. Persisted introspection would probably solve this, and we might be fine with enabling it for a low-traffic collection like this?

Otherwise, adding SUBSCRIBEs to `mz_object_dependencies` would work too, if the surfaces team doesn't think this table should only include catalog objects.

Reply:
> So we should be fine to query any replica for dependency information.

Yes, but if we have several clusters, we will still have to make several queries (one for each cluster) although indeed the replica we choose is arbitrary. We will also have to first query `mz_replicas` to get the list of replicas. So it does make things a bit more complicated.

Another possibility: we don't need to change `mz_object_dependencies`; we could add our own Compute-owned `mz_subscribes` table, as you suggested on the other PR.

Reply:
I *think* that having dataflow dependencies be a controller-level collection makes a lot of sense. I think it commits to instantiating these specific things, with these specific dependencies, to its clients. Even if it does crazy per-replica stuff, we could (imo should) present this information back consistently from the controller.

Rather than derive this from replicas (what if a cluster has no replicas?) I think we should just directly have the controller report it.

Reply:
> Yes, but if we have several clusters, we will still have to make several queries 

Right, that would be annoying.

> I think that having dataflow dependencies be a controller-level collection makes a lot of sense. 

This is already what we do to get a global view of the frontier information, so doing the same thing for dependency information would only make sense!

I wonder if this would make the replica-exported introspection collections redundant, since they contain the same information, just harder to reach.